---
title: "`scfind` package vignette"
author: "Vladimir Kiselev and Jimmy Tsz Hang Lee"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
vignette: >
  %\VignetteIndexEntry{`scfind` package vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE, cache=TRUE}
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png')
op <- options(gvis.plot.tag='chart')
```

# Introduction

The number of cell atlases is growing rapidly, as a result of advances in single-cell sequencing techniques and cheaper sequencing cost. To search through these large single-cell datasets for analysis, however, is time-consuming and inefficient because these types of dataset usually take up large amounts of memory. `scfind` has adopted an efficient compression strategy which makes it suitable for real-time queries of millions of cells.

`scfind` is a method for searching specific cell types from large single-cell datasets by a query of gene list, in which `scfind` can suggest subquries score by TF-IDF method. `scfind` can perform hypergeometric test which allows the evaluation of marker genes specific to each cell type within a dataset.

# `SingleCellExperiment` class

`scfind` is built on top of the Bioconductorâ€™s [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment) class. `scfind` operates on objects of class `SingleCellExperiment` and writes all of its results back to an object. Please read corresponding vignettes on how to create a `SingleCellExperiment` from your own data. Here we will show a small example on how to do that but note that it is not a comprehensive guide. 

# `scfind` Input

If you already have an `SingleCellExperiment` object, then proceed to the next chapter.

If you have a matrix or a data frame containing expression data then you first need to create a `SingleCellExperiment` object containing your data. For illustrative purposes we will use an example expression matrix provided with `scfind`. The dataset (`Tabula Muris (FACS)`) represents __CPM__ gene expression of 517 and 4897 cells derived from mouse kidney and marrow, respectively. The investigators ([The Tabula Muris Consortium](https://doi.org/10.1038/s41586-018-0590-4)) have profiled almost every cell-type in the mouse using high-coverage FACS-sorted cells + Smartseq2. in the original publication. We will combine the indices of `kidney` and `marrow` into a super-index later

```{r , warning=FALSE, message=FALSE}
data(kidney)
data(tabulamuris)

library(SingleCellExperiment)
library(scfind)

head(kidney)
tabulamuris[1:3, 1:3]
```

Note that the cell type information has to be stored in the `cell_type1` column of the `rowData` slot of the `SingleCellExperiment` object.

Now let's create a `SingleCellExperiment` object of the `tabulamuris` dataset:
```{r}
# remove spike-ins
tabulamuris <- tabulamuris[-(grep("ERCC-", rownames(tabulamuris))),]
sce <- SingleCellExperiment(assays = list(normcounts = as.matrix(tabulamuris)), colData = kidney)
# this is needed to calculate dropout rate for feature selection
# important: normcounts have the same zeros as raw counts (cpm)
counts(sce) <- normcounts(sce)
logcounts(sce) <- log2(normcounts(sce) + 1)
# use gene names as feature symbols
rowData(sce)$feature_symbol <- rownames(sce)
isSpike(sce, "ERCC") <- grepl("^ERCC-", rownames(sce))
# remove features with duplicated names
sce <- sce[!duplicated(rownames(sce)), ]
sce
```

# Cell Type Search

## Index

Once we have a `SingleCellExperiment` object, we can run `scfind`. Firstly, we need to build the `scfind` index from our input dataset.

By default `scfind` uses the `cell_type1` column of the `colData` slot in the reference to identify cell type names. 

```{r}
kidneyIndex <- buildCellTypeIndex(sce, "Kidney")

# if the dataset contains more than one category, users can build index for each tissue individually and merge all indices into 1 super-index using `mergeDataset` as following:
data(sce_tmfacs)
marrowIndex <- buildCellTypeIndex(sce.marrow, "Bone Marrow")
geneIndex <- mergeDataset(kidneyIndex, marrowIndex)

geneIndex
# The `scfind` index can be saved as a RDS object 
saveObject(geneIndex, file = "scfindIndex.rds")
```

Once the `scfind` index is built, one can view all existing genes in the datasets using `scfindGenes`

```{r}
head(scfindGenes(geneIndex))
```

or view all existing cell type names in the database using `cellTypeNames`

```{r}
cellTypeNames(geneIndex)
```


## Evaluates Marker Genes

Once we have a `scfind` index, we can find the cell types that most likely represent your list of genes (highest expression) from your dataset very quickly:
```{r}
geneIndex <- loadObject(file = "scfindIndex.rds")

initialList <- c("E230029C05Rik", "Kynu", "Rab5b", "Nedd4l", "Slc24a3", "D630033O11Rik")

scoreGenes <- markerGenes(geneIndex, initialList)
scoreGenes
barplot(scoreGenes$tfidf, main = "TF-IDF score", names.arg = as.character(scoreGenes$Query), ylab="TF-IDF score", cex.names = 0.65, las = 2)
barplot(scoreGenes$Cells, main = "No. of cells", names.arg = as.character(scoreGenes$Query), ylab="no. of cells", cex.names = 0.65, las = 2)
# In which, `scfind` will suggest subqueries for the initial gene sets along with number of cells and cell types that express the genes

bestQuery <- strsplit(as.character(scoreGenes$Query[tail(order(scoreGenes$Genes),1)]), ',')[[1]]
bestQuery
# By ranking the tfidf score, the best subquery can be used to search for cell types that is enriched. The result is reflected by the p-value

enrichedCellTypes <- hyperQueryCellTypes(geneIndex, bestQuery)
enrichedCellTypes
barplot(enrichedCellTypes$cell_hits, main = "Enriched cell types", beside=T, names.arg = gsub("Bone Marrow.", "", enrichedCellTypes$cell_type), ylab="Cell hits",  cex.names = 0.65, las=2)
barplot(-log10(enrichedCellTypes$pval), names.arg = enrichedCellTypes$cell_type, cex.names = 0.65,  ylab = "-log10(pval)", las = 2)
# The calculation above shows that a list of genes containing `Kynu`, `Nedd4l` and `Rab5b` is specific for the `Bone Marrow.Fraction A pre-pro B cell` cell type with the lowest p-value and the highest number of cell hits and total cells.

# To further evaluate a specific query by calculating the precision recall metrics
evaluateGenes <- evaluateMarkers(geneIndex, bestQuery, "Bone Marrow.Fraction A pre-pro B cell")
evaluateGenes
data1 <- t(evaluateGenes[, c(7,6,8)])
colnames(data1) <- evaluateGenes$genes
barplot(data1 , main = paste("Evaluation of markers in Fraction A pre-pro B cell"), beside=T, legend=rownames(data1), ylab="Arbitrary unit", ylim = c(0, .6))

```



# Marker Gene Search

If one is more interested in finding out which marker genes best represent a cell type in the dataset, `cellTypeMarkers` function should be used for searching the index:
```{r}
geneIndex <- loadObject(file = "scfindIndex.rds")
interestedCellType <- c("Bone Marrow.Fraction A pre-pro B cell")
findMarkers <- cellTypeMarkers(geneIndex, interestedCellType)
data2 <- t(findMarkers[, c(7,6,8)])
colnames(data2) <- findMarkers$genes
barplot(data2 , main = paste("Marker genes of ", interestedCellType), font.axis=2, beside=T, legend=rownames(data2), ylab="Arbitrary unit", ylim = c(0, 1.1))
```

To find cell types that is associated with a given gene list, one can use `findCellTypes`

```{r}
findCellTypes(geneIndex, c("Kynu", "Nedd4l", "Rab5b"), "Bone Marrow")
```



# sessionInfo()

```{r echo=FALSE}
sessionInfo()
```

